backend/src/main/java/com/benine/backend/video/Stream.java

```diff
@@ -1,57 +1,132 @@
-    openConnection(url);
+    this.url = new URL(streamurl);
+    this.connected = false;
+    this.logger = ServerController.getInstance().getLogger();
+    this.out = new PipedOutputStream();
+    this.pipedInputStream = new PipedInputStream(out);
+
+    Thread streamThread = new Thread() {
+
+      @Override
+      public void run() {
+        process();
+      }
+    };
+
+    streamThread.start();
 
-    this.inputstream = fetchInputStream();
   }
 
   /**
-   * Opens a connection to the stream.
-   * @param streamURL The url of the stream to open the connection to.
-   * @throws IOException If an error occurs opening the connection.
+   * Processes the stream, and maintains the connection.
+   */
+  private void process() {
+    while (!Thread.interrupted()) {
+      if (!connected) {
+        openConnection();
+      } else {
+        try {
+          streamToOutputstream();
+        } catch (IOException e) {
+          connected = false;
+        }
+      }
+    }
+  }
+
+  /**
+   * Reads the stream and adds it to the outputstream of this instance.
+   * This prevents the stream from ending if the connection to the
+   * stream ends, because it does not send the '-1' termination symbol to
+   * the stream.
+   *
+   * @throws IOException If the stream cannot be read.
    */
-  private void openConnection(URL streamURL) throws IOException {
-    URLConnection conn = streamURL.openConnection();
-    conn.setConnectTimeout(5000);
+  private void streamToOutputstream() throws IOException {
+    byte[] bytes = new byte[BUFFER];
+    int bytesRead;
+
+    if ((bytesRead = in.read(bytes)) > 0) {
+      out.write(bytes, 0, bytesRead);
+    } else {
+      throw new IOException();
+    }
+  }
+
+  /**
+   * Opens a connection to the stream and reconnects on failure.
+   */
+  private void openConnection() {
+    try {
+      URLConnection conn = url.openConnection();
+      conn.connect();
+
+      this.connected = true;
+      this.in = conn.getInputStream();
+
+      logger.log("Connected to stream " + url.toString(), LogEvent.Type.INFO);
+    } catch (IOException e) {
+      logger.log("Could not connect to stream " + url.toString()
+              + ", attempting to reestablish.", LogEvent.Type.WARNING);
 
-    this.connection = conn;
+      // Only retry once every 5 seconds.
+      wait(RECONNECT_DELAY);
+    }
   }
 
   /**
-   * Fetches the inputstream from the connection.
-   * @return  An the inputstream of the feed.
-   * @throws IOException If the inputstream cannot be read from the connection.
+   * Halts a thread for a specified duration and logs it to the logger.
+   * @param duration  Duration in milliseconds.
    */
-  private InputStream fetchInputStream() throws IOException {
-    return connection.getInputStream();
+  private void wait(int duration) {
+    try {
+      Thread.sleep(duration);
+    } catch (InterruptedException e) {
+      logger.log("Stream thread interrupted while sleeping", LogEvent.Type.INFO);
```

Comment:
It might be more usefull to also include the stacktrace (exception) when logging this way you can determine why it was interrupted.